File: Linked list/SwappingNodes_in_a_LinkedList.py
#Problelm Link:- https://leetcode.com/problems/swapping-nodes-in-a-linked-list/

"""You are given the head of a linked list, and an integer k.Return the head of the linked 
list after swapping the values of the kth node from the beginning and the kth node from the 
end"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def swapNodes(self, head: ListNode, k: int) -> ListNode:
        if head.next == None:
            return head
        slow,fast = head,head
        for i in range(k-1):
            fast = fast.next
        first = fast
        while fast.next:
            fast = fast.next
            slow =  slow.next
        second = slow
        first.val, second.val = second.val,first.val
        return head

File: Linked list/SwappingNodes_in_a_LinkedList.cpp
//Problem Link:- https://leetcode.com/problems/swapping-nodes-in-a-linked-list/

#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        if (head == NULL){
            return head;
        }
        if (head->next == NULL){
            return head;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        int cnt = 0;
        while (cnt != k && fast != NULL){
            fast = fast->next;
            cnt++;
        }
        while (fast != NULL){
            fast = fast->next;
            slow = slow->next;
        }
        cnt = 1;
        fast = head;
        while (cnt != k && fast != NULL){
            fast = fast->next;
            cnt++;
        }
        int temp = fast->val;
        fast->val = slow->val;
        slow->val = temp;
        return head;
    }
};